src/from_fn.rs

```diff
@@ -0,0 +1,132 @@
+    where
+        F: FnMut(usize) -> T,
+    {
+        Array::from_fn(cb)
+    }
+
+    #[inline]
+    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>,
+    {
+        Array::try_from_fn(cb)
+    }
+}
+
+impl<T, const N: usize> FromFn<T> for [T; N] {
+    #[inline]
+    fn from_fn<F>(cb: F) -> Self
+    where
+        F: FnMut(usize) -> T,
+    {
+        core::array::from_fn(cb)
+    }
+
+    // TODO(tarcieri): use `array::try_from_fn` when stable
+    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>,
+    {
+        // SAFETY: an array of `MaybeUninit`s is always valid.
+        let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
+        try_from_fn_erased(&mut array, cb)?;
+
+        // TODO(tarcieri): use `MaybeUninit::array_assume_init` when stable
+        // SAFETY: if we got here, every element of the array was initialized
+        let res: Self = unsafe { ptr::read(array.as_ptr().cast()) };
+        mem::forget(array);
+        Ok(res)
+    }
+}
+
+/// Fills a `MaybeUninit` slice using the given fallible generator function.
+///
+/// Using a slice avoids monomorphizing for each array size.
+#[inline]
+fn try_from_fn_erased<T, E, F>(buffer: &mut [MaybeUninit<T>], mut cb: F) -> Result<(), E>
+where
+    F: FnMut(usize) -> Result<T, E>,
+{
+    let mut guard = Guard {
+        array_mut: buffer,
+        initialized: 0,
+    };
+
+    while guard.initialized < guard.array_mut.len() {
+        let item = cb(guard.initialized)?;
+
+        // SAFETY: the loop's condition ensures we won't push too many items
+        unsafe { guard.push_unchecked(item) };
+    }
+
+    mem::forget(guard);
+    Ok(())
+}
+
+/// Drop guard which tracks the total number of initialized items, and handles dropping them in
+/// the event a panic occurs.
+///
+/// Use `mem::forget` when the array has been fully constructed.
+struct Guard<'a, T> {
+    /// Array being constructed.
+    array_mut: &'a mut [MaybeUninit<T>],
+
+    /// Number of items in the array which have been initialized.
+    initialized: usize,
+}
+
+impl<T> Guard<'_, T> {
+    /// Push an item onto the guard, writing to its `MaybeUninit` slot and incrementing the
+    /// counter of the number of initialized items.
+    ///
+    /// # Safety
+    ///
+    /// This can only be called n-times for as many elements are in the slice.
+    #[inline]
+    pub unsafe fn push_unchecked(&mut self, item: T) {
+        unsafe {
+            self.array_mut
+                .get_unchecked_mut(self.initialized)
+                .write(item);
+            self.initialized = self.initialized.saturating_add(1);
+        }
+    }
+}
+
+impl<T> Drop for Guard<'_, T> {
+    fn drop(&mut self) {
+        debug_assert!(self.initialized <= self.array_mut.len());
+
+        unsafe {
+            crate::ptr::drop_in_place(self.array_mut.get_unchecked_mut(..self.initialized));
```

Comment:
I don't think it works as you think it is, since you try to call `Drop` on `MaybeUninit`. We probably need something like this:
```rust
let p: *mut T = self.array_mut.as_mut_ptr().cast();
for i in 0..self.initialized {
    unsafe { ptr::drop_in_place(p.add(i)); }
}
```

Reply:
Aah yes indeed, we need a proper replacement for `MaybeUninit::slice_assume_init_mut`

Reply:
Fixed in [8a89f17](https://github.com/RustCrypto/hybrid-array/pull/2/commits/8a89f1701ab59681bc65ba3b6f32ca0cd3dbd883)